### PR TITLE
12 move and update align timestamps

### PR DESCRIPTION
**Pull request by njmei**
This PR is intended to address issue #12.

This PR should be reviewed *AFTER* #16, but should be merged *BEFORE* it.

Along the way of getting tests to pass and new VBN data to work properly, it was discovered that old tests were using on-prem data resources. The brunt of the work in this PR was developing more light weight test resources for both Visual Coding Neuropixels as well as new test resources for Visual Behavior Neuropixels for testing `align_timestamps` functionality.

### CODE REVIEW

File: src/ecephys_etl/modules/align_timestamps/_schemas.py [hunk 1]
```diff
@@ -10,40 +10,62 @@ class ProbeMappable(DefaultSchema):
     )
     input_path = String(
```
Comment (danielsf) [inline src/ecephys_etl/modules/align_timestamps/_schemas.py:11]: Please make this an `argschema.Fields.InputFile`

File: src/ecephys_etl/modules/align_timestamps/_schemas.py [hunk 2]
```diff
@@ -10,40 +10,62 @@ class ProbeMappable(DefaultSchema):
...
         required=True,
-        help="Input path for this file. Should point to a file containing a 1D timestamps array with values in probe samples.",
+        help=(
+            "Input path for this file. Should point to a file containing "
+            "a 1D timestamps array with values in probe samples."
+        ),
     )
     output_path = String(
```
Comment (danielsf) [inline src/ecephys_etl/modules/align_timestamps/_schemas.py:18]: Please make this an `argschema.Fields.OutputFile`

File: src/ecephys_etl/modules/align_timestamps/_schemas.py [hunk 3]
```diff
@@ -10,40 +10,62 @@ class ProbeMappable(DefaultSchema):
...
-        required=True, help="The sampling rate of the LFP collected on this probe."
+        required=True,
+        help="The sampling rate of the LFP collected on this probe."
     )
     start_index = Int(
-        default=0, help="Sample index of probe recording start time. Defaults to 0."
+        default=0,
+        help="Sample index of probe recording start time. Defaults to 0."
     )
     barcode_channel_states_path = String(
```
Comment (danielsf) [inline src/ecephys_etl/modules/align_timestamps/_schemas.py:44]: Please make this an `argschema.Fields.InputFile`

File: src/ecephys_etl/modules/align_timestamps/_schemas.py [hunk 4]
```diff
@@ -10,40 +10,62 @@ class ProbeMappable(DefaultSchema):
...
-        help="Path to the channel states file. This file contains a 1-dimensional array whose axis is events and whose "
-        "values indicate the state of the channel line (rising or falling) at that event.",
+        help=(
+            "Path to the channel states file. This file contains a "
+            "1-dimensional array whose axis is events and whose "
+            "values indicate the state of the channel line "
+            "(rising or falling) at that event."
+        ),
     )
     barcode_timestamps_path = String(
```
Comment (danielsf) [inline src/ecephys_etl/modules/align_timestamps/_schemas.py:53]: Please make this an `argschema.Fields.InputFile`

File: src/ecephys_etl/modules/align_timestamps/_schemas.py [hunk 5]
```diff
@@ -54,7 +76,8 @@ class InputParameters(ArgSchema):
         help="Probes whose data will be aligned to the master clock.",
     )
     sync_h5_path = String(
```
Comment (danielsf) [inline src/ecephys_etl/modules/align_timestamps/_schemas.py:78]: Please make this an `argschema.Fields.InputFile`

File: src/ecephys_etl/modules/align_timestamps/barcode.py [hunk 6]
```diff
@@ -4,10 +4,10 @@
 def extract_barcodes_from_times(
     on_times,
```
Comment (danielsf) [inline src/ecephys_etl/modules/align_timestamps/barcode.py:5]: Please add type hinting for on_times and off_times

File: src/ecephys_etl/modules/align_timestamps/barcode.py [hunk 7]
```diff
@@ -50,12 +49,16 @@ def extract_barcodes_from_times(
 
         oncode = on_times[
             np.where(
```
Comment (danielsf) [inline src/ecephys_etl/modules/align_timestamps/barcode.py:51]: Do you need the `np.where` here? Can we just use the boolean mask?

File: src/ecephys_etl/modules/align_timestamps/barcode.py [hunk 8]
```diff
@@ -50,12 +49,16 @@ def extract_barcodes_from_times(
...
-                np.logical_and(on_times > t, on_times < t + barcode_duration_ceiling)
+                np.logical_and(
+                    on_times > t, on_times < t + barcode_duration_ceiling
+                )
             )[0]
         ]
         offcode = off_times[
             np.where(
```
Comment (danielsf) [inline src/ecephys_etl/modules/align_timestamps/barcode.py:58]: Do you need the `np.where` here? Can we just use the boolean mask?

File: src/ecephys_etl/modules/align_timestamps/barcode.py [hunk 9]
```diff
@@ -94,10 +97,12 @@ def extract_barcodes_from_times(
     return barcode_start_times, barcodes
 
 
-def find_matching_index(master_barcodes, probe_barcodes, alignment_type="start"):
-    """Given a set of barcodes for the master clock and the probe clock, find the
-    indices of a matching set, either starting from the beginning or the end
-    of the list.
+def find_matching_index(
+    master_barcodes, probe_barcodes, alignment_type="start"
```
Comment (danielsf) [inline src/ecephys_etl/modules/align_timestamps/barcode.py:101]: Please add type hinting to this call signature

File: src/ecephys_etl/modules/align_timestamps/barcode.py [hunk 10]
```diff
@@ -147,34 +150,35 @@ def find_matching_index(master_barcodes, probe_barcodes, alignment_type="start")
 
 
 def match_barcodes(master_times, master_barcodes, probe_times, probe_barcodes):
```
Comment (danielsf) [inline src/ecephys_etl/modules/align_timestamps/barcode.py:152]: Please add type hinting to this call signature

File: src/ecephys_etl/modules/align_timestamps/channel_states.py [hunk 11]
```diff
@@ -16,12 +16,11 @@ def extract_barcodes_from_states(
         Sample index of each event.
     sampling_rate : numeric
         Samples / second
-    **barcode_kwargs : 
+    **barcode_kwargs :
```
Comment (danielsf) [inline src/ecephys_etl/modules/align_timestamps/channel_states.py:19]: Can we at least refer the user to a different docstring that explicitly defines the allowed/expected kwargs?

File: src/ecephys_etl/modules/align_timestamps/channel_states.py [hunk 12]
```diff
@@ -44,12 +43,11 @@ def extract_splits_from_states(
...
```
Comment (danielsf) [inline src/ecephys_etl/modules/align_timestamps/channel_states.py:46]: Can we refer to a different docstring that explicitly defines the expected kwargs?

File: src/ecephys_etl/modules/align_timestamps/probe_synchronizer.py [hunk 13]
```diff
@@ -122,32 +126,37 @@ def compute(
...
+            the probe clock.
 
         """
 
         times_array = np.array(probe_barcode_times)
         barcodes_array = np.array(probe_barcodes)
 
-        ok_barcodes = np.where((times_array > min_time) * (times_array < max_time))[0]
+        ok_barcodes = np.where(
+            (times_array > min_time) * (times_array < max_time)
```
Comment (danielsf) [inline src/ecephys_etl/modules/align_timestamps/probe_synchronizer.py:142]: Can we make this an `np.logical_and`. Multiplying booleans this way, while legal, is confusing to me.

**Review by danielsf — CHANGES_REQUESTED** [review]
Minor changes requested

File: src/ecephys_etl/modules/align_timestamps/README.md [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+- For each probe
+    - barcode channel states file: lists rising and falling edges on the probe's barcode line
+    - barcode timestamps file: lists probe samples at which rising and falling edges were detected
+    - mappable timestamp files: Will be transformed to the master clock. An example would be a file listing timestamps of detected spikes.
+
+
+
+Output data
+-----------
+Each mappable file for each probe is aligned and written out. Additionally, the transform is written into the output json.
```
Comment (djkapner) [inline src/ecephys_etl/modules/align_timestamps/README.md:27]: this is a barely useful amount of information.

File: src/ecephys_etl/modules/align_timestamps/__main__.py [hunk 15]
```diff
@@ -1,23 +1,59 @@
...
+from typing import Dict, List, TypedDict
 
+import numpy as np
 
-def align_timestamps(args):
-    sync_dataset = BarcodeSyncDataset.factory(args["sync_h5_path"])
+from argschema import ArgSchemaParser
+
+from ecephys_etl.modules.align_timestamps._schemas import (
+    InputParameters, OutputParameters
```
Comment (djkapner) [inline src/ecephys_etl/modules/align_timestamps/__main__.py:8]: would be nice to have more specific names, for github searchability, grep, etc.

File: src/ecephys_etl/modules/align_timestamps/_schemas.py [hunk 16]
```diff
@@ -10,40 +10,62 @@ class ProbeMappable(DefaultSchema):
...
-        help="Input path for this file. Should point to a file containing a 1D timestamps array with values in probe samples.",
+        help=(
+            "Input path for this file. Should point to a file containing "
+            "a 1D timestamps array with values in probe samples."
+        ),
     )
     output_path = String(
         required=True,
-        help="Output path for the mapped version of this file. Will write a 1D timestamps array with values in seconds on the master clock.",
+        help=(
```
Comment (djkapner) [inline src/ecephys_etl/modules/align_timestamps/_schemas.py:20]: `help` ->  `description` would be consistent with everywhere else

File: src/ecephys_etl/modules/align_timestamps/channel_states.py (hunk 11 above)
Comment (djkapner) [reply src/ecephys_etl/modules/align_timestamps/channel_states.py:19]: I agree, I think this needs to be explicit, or typed out somehow.

**Comment by njmei** [comment]
Branch has been updated to address comments. Module has been run with dev LIMS and looks to be working:

image

**Review by djkapner — APPROVED** [review]

File: tests/modules/align_timestamps/test_align_timestamps_module.py [hunk 17]
```diff
@@ -1,186 +1,133 @@
...
-        )
-        assert (
-            expected["global_probe_lfp_sampling_rate"]
-            == probe["global_probe_lfp_sampling_rate"]
-        )
+], indirect=["align_timestamps_test_case_fixture"])
+def test_align_timestamps(align_timestamps_test_case_fixture):
+    test_input, test_expected = align_timestamps_test_case_fixture
 
+    print(test_input)
```
Comment (danielsf) [inline tests/modules/align_timestamps/test_align_timestamps_module.py:116]: Should these print statements be here?

**Review by danielsf — APPROVED** [review]
Looks like some print statements got left in a test by accident. Otherwise: looks good

File: tests/modules/align_timestamps/test_align_timestamps_module.py (hunk 17 above)
Comment (njmei) [reply tests/modules/align_timestamps/test_align_timestamps_module.py:116]: I'm generally okay with print statements in test code. Pytest suppresses any print statements such that they only show up when their associated test fails.

**Pull request merged**